modules/lib/lib-portal/src/main/resources/lib/xp/portal.ts

```diff
@@ -83,6 +83,86 @@ export interface SiteConfig<Config> {
     config: Config;
 }
 
+export type DefaultHeaders = {
+    Accept: string;
+    'Accept-Charset': string;
+    'Accept-Encoding': string;
+    Authorization: string;
+    'Content-Length': string,
+    'Content-Type': string,
+    Cookies: string;
+    Expect: string;
+    Host: string;
+    'If-None-Match': string;
+    Language: string;
+    'User-Agent': string;
+};
+
+export interface Cookie {
+    value: string;
+    path?: string;
+    domain?: string;
+    comment?: string;
+    maxAge?: number; // -1
+    secure?: boolean; // false
+    httpOnly?: boolean; // false
+    sameSite?: 'lax' | 'strict' | 'none' | ''; // ''
+}
+
+interface Request<Params extends Record<string, string | undefined> = Record<string, string | undefined>,
+    Headers extends Record<string, string | undefined> = DefaultHeaders,
+    Cookies extends Record<string, string | Cookie | undefined> = Record<string, string | Cookie | undefined>,
+    > {
+    method: 'GET' | 'PUT' | 'POST' | 'DELETE' | 'HEAD' | 'PATCH' | 'OPTIONS' | 'TRACE' | 'CONNECT';
+    scheme: 'http' | 'https';
+    host: string;
+    port: number;
+    path: string;
+    rawPath: string;
+    url: string;
+    remoteAddress: string;
+    mode: 'inline' | 'edit' | 'preview' | 'live';
+    webSocket?: boolean;
+    repositoryId: string;
+    branch: 'draft' | 'master';
+    contextPath: string;
+    body: string;
+    params: Params;
+    headers: Headers;
+    cookies: Cookies;
+    contentType: string;
+}
+
+type ResponseType = string | object | unknown[] | ReadonlyArray<unknown> | null;
+
+type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>);
+
+export interface PageContributions {
+    headBegin?: string | string[];
+    headEnd?: string | string[];
+    bodyBegin?: string | string[];
+    bodyEnd?: string | string[];
+}
+
+export interface Response<ResponseBody extends ResponseType = ResponseType,
+    Headers extends Record<string, string | undefined> = DefaultHeaders,
+    Cookies extends Record<string, string | Cookie | undefined> = Record<string, string | Cookie | undefined>,
+    > {
+    status: number;
+    body: ResponseBody;
+    contentType: LiteralUnion<| 'text/html'
```

Comment:
I am surprised with a set of predefined ones. I suggest to leave just string

Reply:
I am ambivalent about this one. :)

**Pro:** Since it's a `LiteralUnion` it can be any `string`. But I get that nice code completion for the most common cases, so that I don't have to google the syntax every time (and trust me, I will google it every time to be sure).

**Con:** It can be anything, so why limit it? And this doesn't really have anything to do with XP.

Reply:
It's just a suggested values, `contentType` can be any string, but those one are the most used and therefore listed here.

Reply:
Content-type is a tricky one and error prone.
The suggested "most used" are not without problems either.
For instance : `text/html` without charset may end up not decoding correctly in browsers.

Reply:
I see. Then let us use the `string` here